### PR TITLE
Working on some performance (GPU utilization/memory, img/sec) gains

### PR DESCRIPTION
**Pull request by rwightman**

### CODE REVIEW

**Comment by bonlime** [comment]
Hi, I've noticed this branch and want to give some small hints.
FIrst of all thanks for such amazing repo, I've learned a lot looking through your code and even borrowed some parts :) 
I've been working on optimizing EffDet training for a couple of days now and this is what I've got.
1. The low hanging fruits are optimizing functions which work with bbox to also work with batch dimension(`box_coder.py`, `argmax_matcher.py`, `box_list.py` in your repo. I already have working and covered with tests versions [in a standalone file here](https://github.com/bonlime/pytorch-tools/blob/effdet_dev/pytorch_tools/utils/box.py). You could take a look.
The largest improvement came from eliminating loop in target assigner
2. https://github.com/rwightman/efficientdet-pytorch/blob/f605bf2c3cf97cde6b19865d052548c4506ca24b/effdet/anchors.py#L390
It is much faster to remove all loops. This could be done by concatenation of all outputs into one large tensor as early as possible. In my implementation I follow https://github.com/NVIDIA/retinanet-examples and return from model already flattened boxes and classes. For input 512x512 box shape is `(5, 49104, 4)` and class shape is `(5, 49104, 90)`. Anchors should be flattened in the same order as outputs.
3. Focal loss computation also requires a lot of memory. Using JIT script helped me to reduce the footprint by a lot
```
JIT Script for Focal
Mean of 10 runs 10 iters each BS=16, SZ=512:
   78.00+-1.95 msecs Forward. 186.96+-4.82 msecs Backward. Max memory: 6929.30Mb. 60.39 imgs/sec
No JIT
Mean of 10 runs 10 iters each BS=16, SZ=512:
   88.07+-1.04 msecs Forward. 186.38+-7.90 msecs Backward. Max memory: 7738.79Mb. 58.30 imgs/sec
```

As for now, I'm able to get ~60 imgs/s for EffDetB0 on V100 with apex O1. Your code from master gives ~40 imgs/s on my machine.

**Comment by rwightman** [comment]
@bonlime thanks for the links, I'll take a look as I add to the opts, definitely lots of room for improvement, especially for training. 

I tried to jit the full DetectionLoss forward when I did some refactoring for jit compatibility it destabilized training and blew up so I haven't revisited. Did you just jit the focal loss, also box?

**Comment by rwightman** [comment]
looks like torch script conflicting with amp's auto casting... I can work around it by forcing some of the tensors in the loss to fp32.

**Comment by bonlime** [comment]
You're right that jitting destabilizes training. It was fine at first but stopped converging quickly. I'm trying to figure out where exactly the problem occurs

**Comment by bonlime** [comment]
So I've moved to the nightly 1.6 build of PyTorch and changed apex.amp to the native amp. It requires very minimal changes to the code, works out of the box and gives exactly the same speed for not scripted loss. But what is important - scripting whole detention loss works perfectly fine, without any need for patching tensors to fp32 manually.
